prow/pubsub/subscriber/subscriber.go

```diff
@@ -154,16 +172,70 @@ func (peh *periodicJobHandler) getProwJobSpec(cfg *config.Config, pe ProwJobEven
 }
 
 // presubmitJobHandler implements jobHandler
-type presubmitJobHandler struct{}
+type presubmitJobHandler struct {
+	GitHubClient githubClient
+	GitClient    git.ClientFactory
+}
+
+func (prh *presubmitJobHandler) getProwJobSpec(cfg prowCfgClient, pe ProwJobEvent) (*v1.ProwJobSpec, map[string]string, error) {
+	// presubmit jobs require Refs and Refs.Pulls to be set
+	refs := pe.Refs
+	if refs == nil {
+		return nil, nil, errors.New("Refs must be supplied")
+	}
+	if len(refs.Pulls) != 1 {
+		return nil, nil, fmt.Errorf("expect 1 pull, got: %d", len(refs.Pulls))
+	}
+	pr := refs.Pulls[0]
+
+	var presubmitJob *config.Presubmit
+	org, repo, branch := refs.Org, refs.Repo, pr.Ref
+	orgRepo := org + "/" + repo
+	baseSHAGetter := func() (string, error) {
+		if refs.BaseSHA != "" {
+			return refs.BaseSHA, nil
+		}
+		// Fall back to fetch from GitHub
+		baseSHA, err := prh.GitHubClient.GetRef(refs.Org, refs.Repo, "heads/"+refs.BaseRef)
```

Comment:
I'm not sure how I feel about this. On one hand it helps ensure that the presubmit spec is deterministic by ensuring the baseref is resolved once, rather than at runtime every time it is rerun. On the other hand this assumes the code review system is GitHub. This only plays a role if the user fails to supply the base SHA in their pubsub request so I don't feel too strongly, but I'd lean toward removing this assumption and just omit the BaseSHA if it is not specified in the request. WDYT?

Reply:
Totally makes sense, deferred to runtime